# 01 逻辑运算

# 02 三元运算

# 03 一行代码实现数值交换

# 04 字符串反转

# 05 基本数据类型False值

# 06 is和==

# 07互不相同的三位数
1,2,3,4,5 能相合成多少个互不相同且无重重的三位數

In [7]:
def calc():
    rst = []
    for i in range(100, 1000):
        li = [int(j) for j in str(i)]
        s = set(li)
        if len(s) == 3 and s & {1, 2, 3, 4, 5} == s:
            rst.append(li[0] * 100 + li[1] * 10 + li[2])
    print(rst)
    return len(rst)


calc()


[123, 124, 125, 132, 134, 135, 142, 143, 145, 152, 153, 154, 213, 214, 215, 231, 234, 235, 241, 243, 245, 251, 253, 254, 312, 314, 315, 321, 324, 325, 341, 342, 345, 351, 352, 354, 412, 413, 415, 421, 423, 425, 431, 432, 435, 451, 452, 453, 512, 513, 514, 521, 523, 524, 531, 532, 534, 541, 542, 543]


60

# 08 fromkeys

In [8]:
a = dict.fromkeys(['v1', 'v2'], [])

# 09 一行代码删除列表重复的值

In [10]:
set([1, 2, 3, 4, 5, 6, 3, 3, 1])

{1, 2, 3, 4, 5, 6}

# 13 合并算法

In [12]:
a = [
    'a,1',
    'b,3,22',
    'c,3,4'
]
b = [
    'a,2',
    'b,1',
    'd,2'
]

for i in range(len(b)):
    for j in range(len(a)):
        if b[i][0] == a[j][0]:
            a[j] = a[j] + b[i][1:]
            break
    else:
        a.append(b[i])
print(a)

['a,1,2', 'b,3,22,1', 'c,3,4', 'd,2']


# 14 extend

In [18]:
a = [1, 2, [3, 4, [5, 6]]]
c = []


def chain(li):
    for item in a:
        if isinstance(item, list):
            c.extend(chain(item))
        else:
            c.append(item)


chain(a)

print(c)

RecursionError: maximum recursion depth exceeded

# 15
将列表内的元素，根图位豉合并成字典
例：
lst = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536, 4294967296]
{1: [1, 2, 4, 8],
 2: [16, 32, 64],
 3: [128, 256, 512],
 4: [1024, 2048, 4096, 8192],
 5: [16384, 32768, 65536],
 10: [4294967296]}

In [19]:
from collections import defaultdict

lst = [1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536, 4294967296]
a = defaultdict(list)
for i in lst:
    a[len(str(i))].append(i)
print(dict(a))


{1: [1, 2, 4, 8], 2: [16, 32, 64], 3: [128, 256, 512], 4: [1024, 2048, 4096, 8192], 5: [16384, 32768, 65536], 10: [4294967296]}


# 16 二维数组转一维数组
轉換前
Lst
[[1,2,31, [4,5,61, 07,8,91]
轉换后
lst = 11,2,3,4,5,6,7,8,9]

In [22]:
sum([[1, 2, 3], [4, 5, 6], [7, 8, 9]], [])

[1, 2, 3, 4, 5, 6, 7, 8, 9]

# 17 列表排序

In [24]:
a = [1, 2, 3, 5, 7, 43, 1, 6, 3, 7, 23]
a.sort()
print(a)

[1, 1, 2, 3, 3, 5, 6, 7, 7, 23, 43]


# 26 字符串压缩小算法
把字符串aaabbcccd这行形式的字宇符串压缩成a3b2c3d1这🀄️形式

In [31]:
source_str = 'aaabbcccd'
result_str = ''
f = source_str[0]
cnt = 1
for ch in source_str:
    if ch == f:
        cnt += 1
        continue
    result_str += f'{f}{str(cnt)}'
    f = ch
    cnt = 1
result_str += f'{f}{str(cnt)}'
print(result_str)

a4b2c3d1


# 完数
一个数如果恰好等于它的因子之和，这个数就称为完数、例如：6=1+2+3. 编程找出1000以肉的所有完酸

In [41]:
def yingzi(n):
    y = []
    for i in range(1, (n // 2) + 1):
        if n % i == 0:
            y.append(i)
    return y


def wanshu():
    i = 1
    while True:
        if i == sum(yingzi((i))):
            yield i
        i += 1


qq = wanshu()
print(next(qq))
print(next(qq))
print(next(qq))

6
28
496


# 28
输入一个字行串，输出该字行串中字符的所有組合。
例如：
输入：1，2，3
翰出：1,2,3, 12, 13, 23, 123（ 组合数，不考虑顺序，所以12和21是等价的）

求子集

In [43]:
def subsets(nums):
    res = [[]]
    for num in nums:
        res += [i + [num] for i in res]
    return [i for i in res if i]


subsets([1, 2, 3])

[[1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]

# 29

In [45]:
def get_sub(s):
    sub = ''
    for i in s[:len(s) // 2]:
        sub += i
        x = [item for item in s.split(sub) if item]
        if not x:
            return True, sub
    return False, ''


s = 'abcabcabc'
get_sub(s)

(True, 'abc')

# 30

In [47]:
lst = [12, 213, 22, 2, 2, 2, 22, 2, 22, 32]
rst = [v for idx, v in enumerate(lst) if idx % 2 == 1]
print(rst)
list(filter(lambda x: x[0] % 2 == 1, enumerate(lst)))

[213, 2, 2, 2, 32]


[(1, 213), (3, 2), (5, 2), (7, 2), (9, 32)]

# 31 62进制

In [48]:
def base62encode(n):
    s = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    rst = []
    while n > 0:
        y = n % 62
        rst.append(s[y])
        n = n // 62
    return ''.join(rst[::-1])


base62encode(73)

'1B'

# 32 限制函数执行频率
写一个装饰器，限制函数执行频率，如10秒一次

In [50]:
import time


# 写一个装饰器，限制函数执行频率，如10秒一次
def frequent(func):
    start = 0

    def wrapper(*args, **kwargs):
        nonlocal start
        now = time.time()
        if now - start > 10:
            start = now
            return func(*args, **kwargs)
        else:
            print('频繁访问')

    return wrapper


@frequent
def test_func():
    print(1)


@frequent
def test_func1():
    print(2)


test_func()
test_func()
test_func1()
test_func1()

1
频繁访问
2
频繁访问
